New Hibernate integration module

chkal opened the pull request:
This pull requests provides a new module for integration with [Hibernate](http://hibernate.org/orm/). Hibernate provides various statistics using the [Statistics](https://docs.jboss.org/hibernate/orm/5.2/javadocs/org/hibernate/stat/Statistics.html) interface which can be accessed though the `SessionFactory`.

A few notes:
 * The module depends on the oldest Hibernate version which is still supported. IMO this is reasonable to support Hibernate versions which are typically used in real world projects.
 * I followed the "Best Practices" guide regarding the metric & label naming. However, every feedback is welcome.
 * The unit test mocks the Hibernate Statistics interface to check that all metrics are resolved and named correctly.
 * The help texts for the metrics are copied from the corresponding Javadocs of the `Statistics` interface.

Any kind of feedback is welcome.

brian-brazil commented:
> The module depends on the oldest Hibernate version which is still supported. IMO this is reasonable to support Hibernate versions which are typically used in real world projects.

That depends on the support policy. That'd not be sane for Java or Ruby for example. 4.2 came out about 4 years ago, so that should be safe in this case. Supporting only 4.2.21 would be a problem, as that's only 14 months old.

brian-brazil commented on simpleclient_hibernate/pom.xml:
```diff
@@ -0,0 +1,67 @@
+        <dependency>
+            <groupId>org.hibernate</groupId>
+            <artifactId>hibernate-core</artifactId>
+            <version>4.2.21.Final</version>
```
Use an older version to be sure we're compatible all the way back to 4.2.0

chkal replied:
Ok, I'll change that.

brian-brazil commented on simpleclient_hibernate/src/main/java/io/prometheus/client/hibernate/HibernateStatisticsCollector.java:
```diff
@@ -0,0 +1,282 @@
+ * <p>
+ * Usage example:
+ * <pre>
+ * new HibernateStatisticsCollector(sessionFactory).register();
```
Can there be multiple session factories in a jvm?

chkal replied:
Yes. Especially in application servers. Actually I was thinking about adding a label identifying the `SessionFactory` to the metrics. But it looks like there isn't something like a "name" of a `SessionFactory`.

brian-brazil replied:
The user will need to specify the name then when creating the collector.

chkal replied:
I added a name to the constructor. This name is then added as a label.

brian-brazil commented on simpleclient_hibernate/src/main/java/io/prometheus/client/hibernate/HibernateStatisticsCollector.java:
```diff
@@ -0,0 +1,282 @@
+    );
+  }
+
+  private CounterMetricFamily createCounter(String name, String help, double value) {
```
These aren't adding much, inline them.

chkal replied:
Ok! I wasn't sure whether there will be some common label for all metrics. That's why I added this method. But I'll drop it.

brian-brazil commented on simpleclient_hibernate/src/main/java/io/prometheus/client/hibernate/HibernateStatisticsCollector.java:
```diff
@@ -0,0 +1,282 @@
+    return Arrays.<MetricFamilySamples>asList(
+
+        createCounter(
+            "hibernate_session_open_total",
```
opened

"open" implies the current number of open sessions


Similarly for the other counters.

chkal replied:
Ok, I changed the names where it makes sense.

brian-brazil commented on simpleclient_hibernate/src/main/java/io/prometheus/client/hibernate/HibernateStatisticsCollector.java:
```diff
@@ -0,0 +1,282 @@
+
+        createCounter(
+            "hibernate_flush_total",
+            "Get the global number of flush executed by sessions",
```
s/Get /

chkal replied:
Ok, I see. That's due to the fact that I copied the texts from the original javadocs. I'll change that.

brian-brazil commented on simpleclient_hibernate/src/main/java/io/prometheus/client/hibernate/HibernateStatisticsCollector.java:
```diff
@@ -0,0 +1,282 @@
+
+        createCounter(
+            "hibernate_session_open_total",
+            "Global number of sessions opened",
```
It wouldn't hurt to put the function name in the help string, it'd save users debugging a hop through this code.

chkal replied:
You are talking about the method name used to gather the metric value? Something like:

    Global number of sessions opened (getSessionOpenCount)

brian-brazil commented on simpleclient_hibernate/src/main/java/io/prometheus/client/hibernate/HibernateStatisticsCollector.java:
```diff
@@ -0,0 +1,282 @@
+
+        createGauge(
+            "hibernate_query_execution_seconds_max",
+            "Get the time in milliseconds of the slowest query.",
```
Over what time period? We generally drop metrics like this as they're not particularly useful statistically.

chkal replied:
AFAIK this is since the session factory startup. I wasn't sure whether it makes sense to add these when I wrote the code, so I've no problem with dropping them.

chkal commented:
I pushed the changes! Any further feedback welcome! :-)

brian-brazil commented on simpleclient_hibernate/src/main/java/io/prometheus/client/hibernate/HibernateStatisticsCollector.java:
```diff
@@ -0,0 +1,261 @@
+            "Global number of cacheable entities/collections successfully retrieved from the cache",
+            statistics.getSecondLevelCacheHitCount()
+        ),
+        new CounterMetricFamily(
```
if you're going to remove the vertical whitespace here, remove it everywhere

brian-brazil commented on simpleclient_hibernate/src/main/java/io/prometheus/client/hibernate/HibernateStatisticsCollector.java:
```diff
@@ -0,0 +1,261 @@
+
+        new CounterMetricFamily(
+            "hibernate_connect_total",
+            "The global number of connections asked by the sessions",
```
either "asked for" or "requested"

brian-brazil commented on simpleclient_hibernate/src/main/java/io/prometheus/client/hibernate/HibernateStatisticsCollector.java:
```diff
@@ -0,0 +1,261 @@
+
+        new CounterMetricFamily(
+            "hibernate_flushed_total",
+            "The global number of flush executed by sessions",
```
flushes

chkal commented:
I just pushed the following changes:

 * I removed the blank lines
 * The user can now specify a name which is added as a label to the metrics
 * I renamed the metrics as requested
 * I added the Hibernate `Statistics` method names to the help text.

Any more feedback?

brian-brazil commented on simpleclient_hibernate/src/main/java/io/prometheus/client/hibernate/HibernateStatisticsCollector.java:
```diff
@@ -0,0 +1,249 @@
+ */
+public class HibernateStatisticsCollector extends Collector {
+
+  private static final List<String> LABEL_NAMES = Collections.singletonList("name");
```
Name is a bit generic.

Considering this is likely a per-db connection pool, I'd suggest database or pool.

brian-brazil replied:
Okay, if you think that's best.

brian-brazil commented on simpleclient_hibernate/src/main/java/io/prometheus/client/hibernate/HibernateStatisticsCollector.java:
```diff
@@ -0,0 +1,249 @@
+            statistics.getFlushCount()
+        ),
+        createCounter(
+            "hibernate_connection_requested_total",
```
I was talking more about the help text than the metric name.

chkal replied:
Ah, OK... Sorry about that. I'll undo the change.

brian-brazil commented on simpleclient_hibernate/src/main/java/io/prometheus/client/hibernate/HibernateStatisticsCollector.java:
```diff
@@ -0,0 +1,249 @@
+  private static final List<String> LABEL_NAMES = Collections.singletonList("name");
+
+  private final Statistics statistics;
+  private final String name;
```
Thinking a bit this isn't going to work with 2 collectors, as they're both outputting the same metric families.

What we probably need is that there's one collector that can handle multiple factories, so in the simple case it'll just have one and then there's another function to add more as needed. You'd then have to iterate through them at collection time.

chkal replied:
Ah, OK. I wasn't aware that different collectors cannot produce the same metric name. But I just found the corresponding code in the registry.

Is there any specific reason why this isn't supported? I could imagine that if a single collector is responsible for monitoring "a thing" (like a DB connection pool), that it would make sense that a collector monitors just a single instance and that users can register the collector multiple times with different configuration. In this case the metric name would of cause be the same and the collector instances would just add different labels. Is there any reason why this isn't supported?

However, if you don't want to change the current behavior, I could implement something like:

```java
new HibernateStatisticsCollector()
    .add(sessionFactory1, "unit1")
    .add(sessionFactory2, "unit2")
    .register();

brian-brazil replied:
Usually with direct instrumentation this doesn't come up, as you'd be sharing the metrics across instances as class variables. The majority of custom collectors also only monitor one thing, so the Java client keeps all the handling simple.

If you could change your code that'd be great.

chkal replied:
Okay. I'll add support for multiple session factories to the collector. I'll come back to you as soon as this is done.

brian-brazil commented:
The docs change has also been merged in now, so please add an example to the README.md.

chkal commented:
I just pushed the following changes:

 * I fixed the help texts as requested
 * The collector now supports multiple session factories
 * The `name` label has been renamed to `unit`
 * I added a section about the collector to the new README file 

A small note: I had to rebase this branch against master to get the current version of the README.

brian-brazil commented on README.md:
````diff
@@ -244,6 +244,36 @@ To register the Logback collector can be added to the root level like so:
+`SessionFactory` instances. To use this collector add the following dependency:
+
+```xml
+<dependency>
````
We haven't included this with the others, and I'd prefer not to have another place to manually update version numbers.

brian-brazil commented on simpleclient_hibernate/src/main/java/io/prometheus/client/hibernate/HibernateStatisticsCollector.java:
```diff
@@ -0,0 +1,446 @@
+ *
+ * @author Christian Kaltepoth
+ */
+public class HibernateStatisticsCollector extends Collector {
```
I think an explicit constrcutor would be clearer to those reading the code, and a convenience constructor that also did the add would be useful too.

chkal commented:
I've added two constructors and updated the README as requested.

chkal commented:
I just added a minor improvement for the documentation regarding how to use this collector in a JPA environment.

So all your comments are addressed now. Let me know if there is anything left I can do to get this merged.

brian-brazil commented on simpleclient_hibernate/src/main/java/io/prometheus/client/hibernate/HibernateStatisticsCollector.java:
```diff
@@ -0,0 +1,472 @@
+
+  private static final List<String> LABEL_NAMES = Collections.singletonList("unit");
+
+  private final Map<String, SessionFactory> sessionFactories = new LinkedHashMap<String, SessionFactory>();
```
Hmm, let's make this a ConcurrentHashMap so that different threads can add themselves without having to worry about synchronisation.

chkal replied:
Okay, makes sense! I replaced it with a ConcurrentHashMap.

brian-brazil commented:
Thanks!

chkal commented:
Awesome! Thank you!
